core/modules/date/date.field.inc

```diff
@@ -557,3 +677,442 @@ function date_field_settings_form($field, $instance, $has_data) {
   module_load_include('inc', 'date', 'date.admin');
   return _date_field_settings_form($field, $instance, $has_data);
 }
+
+/**
+ * Custom callback to provide the widget settings form.
+ *
+ * @param array $instance
+ *   Field instance definition.
+ *
+ * @return array
+ *   Widget settings form structure.
+ */
+function _date_html5_field_widget_settings_form(array $instance) {
+  $settings = $instance['widget']['settings'];
+  $form = array();
+
+  $form['limits'] = array(
+    '#type' => 'fieldset',
+    '#title' => t('Date input limits'),
+    '#description' => t('Applies to start and end if enabled and is relative to the current date.'),
+  );
+  $form['limits']['mode'] = array(
+    '#type' => 'radios',
+    '#title' => t('Limit base'),
+    '#options' => array(
+      'none' => t('No limit'),
+      'byyear' => t('By year'),
+    ),
+    '#default_value' => $settings['limits']['mode'],
+  );
+  $form['limits']['years_back'] = array(
+    '#type' => 'number',
+    '#title' => t('Starting year'),
+    '#min' => -9999,
+    '#max' => 0,
+    '#default_value' => $settings['limits']['years_back'],
+    '#description' => t('Empty value for no limit.'),
+    '#field_suffix' => t('from current year'),
+    '#states' => array(
+      'visible' => array(
+        ':input[name="instance[widget][settings][limits][mode]"]' => array('value' => 'byyear'),
+      ),
+    ),
+  );
+  $form['limits']['years_forward'] = array(
+    '#type' => 'number',
+    '#title' => t('Ending year'),
+    '#min' => 0,
+    '#max' => 9999,
+    '#default_value' => $settings['limits']['years_forward'],
+    '#description' => t('Empty value for no limit.'),
+    '#field_suffix' => t('from current year'),
+    '#states' => array(
+      'visible' => array(
+        ':input[name="instance[widget][settings][limits][mode]"]' => array('value' => 'byyear'),
+      ),
+    ),
+  );
+
+  $form['no_fieldset'] = array(
+    '#type' => 'checkbox',
+    '#title'  => t('Exclude wrapping fieldset'),
+    '#default_value' => !empty($settings['no_fieldset']),
+    '#description' => t('If using simple inputs such as a single textfield or the date popup, omitting the fieldset may provide simpler styling.'),
+  );
+
+  return $form;
+}
+
+/**
+ * Element validation callback for the HTML5 widget.
+ *
+ * This runs once over the whole value array.
+ *
+ * @param array $element
+ *   Form element structure.
+ * @param array $form_state
+ *   The form state of this form submission.
+ */
+function date_html5_element_validate(array $element, array &$form_state) {
```

Comment:
This function should probably be in `date.element.inc` with the other element callbacks.

Reply:
That also came to my mind, but I wasn't sure. Many thanks for verifying. :+1: